src/audio/smart_amp_maxim.c

```diff
@@ -0,0 +1,719 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Maxim Integrated All rights reserved.
+//
+// Author: Ryan Lee <ryans.lee@maximintegrated.com>
+
+#include <sof/audio/smart_amp_maxim.h>
+#include <sof/trace/trace.h>
+#include <sof/drivers/ipc.h>
+#include <sof/ut.h>
+#include <sof/audio/sof_dsm.h>
+
+static const struct comp_driver comp_smart_amp;
+
+/* 167a961e-8ae4-11ea-89f1-000c29ce1635 */
+DECLARE_SOF_RT_UUID("Maxim DSM", maxim_dsm_comp_uuid, 0x167a961e, 0x8ae4,
+		    0x11ea, 0x89, 0xf1, 0x00, 0x0c, 0x29, 0xce, 0x16, 0x35);
+
+DECLARE_TR_CTX(maxim_dsm_comp_tr, SOF_UUID(maxim_dsm_comp_uuid),
+	       LOG_LEVEL_INFO);
+
+struct smart_amp_data {
+	struct sof_smart_amp_config config;
+	struct comp_model_data model;
+	struct comp_buffer *source_buf; /**< stream source buffer */
+	struct comp_buffer *feedback_buf; /**< feedback source buffer */
+	struct comp_buffer *sink_buf; /**< sink buffer */
+	smart_amp_proc process;
+	uint32_t in_channels;
+	uint32_t out_channels;
+
+	struct sof_dsm_struct_t *sof_dsm_handle;
+};
+
+static inline void *smart_amp_malloc(int size)
+{
+	void *ret = rballoc(0, SOF_MEM_CAPS_RAM, size);
+
+	if (!ret) {
+		rfree(ret);
+		return NULL;
+	}
+	memset(ret, 0, size);
+	return ret;
+}
+
+static inline int smart_amp_alloc_memory(struct smart_amp_data *sad,
+					 struct comp_dev *dev)
+{
+	struct sof_dsm_struct_t *hsof_dsm;
+	int mem_sz[3];
+	int size;
+
+	if (sad->sof_dsm_handle)
```

Comment:
Needed ?

Reply:
It is not necessary if create function never called twice. This seems true. Shall remove this condition.